### PR TITLE
feat(`cheatcodes`): `recordAccountAccesses` and `recordStorageAccesses` cheatcodes

### PR DESCRIPTION
**Pull request by jameswenzel**
## Motivation

Currently, it is not possible to get an ordered list of account or storage accesses made by the EVM during test execution.

For account accesses, there are cheats such as `vm.expectCall`, but this requires knowing the exact destination and parameters of calls ahead of time. 

For storage accesses, there are the `vm.record/vm.accesses` cheats, but these are not enumerated or ordered overall, and since it is a mapping, it require knowing which accounts exactly you'd like to check storage accesses for (this is made easier by the new mapping cheatcodes).

Further, it is useful to know the context of account and storage accesses, in particular:

For accounts
- what account is being accessed?
- is the account access the result of a `create` or a `call` opcode?
- is the account initialized or empty?
- is value passed along with the account access?
- what input data was provided to the `create` or `call`?
- was this access made as part of a context that was reverted?

For storage
- which account's storage was accessed?
- which slot was accessed?
- was this access a read or a write?
- what was the value before this access?
- what was the value after this access?
- was this access made as part of a context that was reverted?

This information would allow for things like manual gas accounting due to accounts and storage slots marked as "warm" during test setup during execution.

## Solution

This PR adds the following cheatcodes and structs:

```solidity

struct AccountAccess {
    address account;
    bool isCreate;
    bool initialized;
    uint256 value;
    bytes data;
    bool reverted;
}

struct StorageAccess {
    address account;
    bytes32 slot;
    bool isWrite;
    bytes32 previousValue;
    bytes32 newValue;
    bool reverted;
}

// Record all account accesses as part of CREATE or CALL opcodes in order,
// along with the context of the calls
function recordAccountAccesses() external;

// Returns an ordered array of all account accesses
function getRecordedAccountAccesses() external returns (AccountAccess[] memory);

// Record all storage accesses in order, along with the context of the accesses
function recordStorageAccesses() external;

// Returns an ordered array of all storage accesses
function getRecordedStorageAccesses() external returns (StorageAccess[] memory);
```

### For accounts

Internally, `Forge` maintains a `Vec<Vec<RecordedAccountAccess>>`. The vector is 2-dimensional to allow for grouping account accesses by relative call depth, so they can be updated with the eventual failure status of that callframe before being merged with the previous `Vec` to preserve absolute ordering.

On each `call` or `create`, a single-element `Vec<RecordedAccountAccess>` is appended to the state with the context of the `CALL` or `CREATE` opcode.

On `call_end` or `create_end`, the last `Vec<RecordedAccountAccess>` is `pop()`'d off and the accesses therein are updated with the failure status of the `CREATE` or `CALL` opcode. This `Vec<RecordedAccountAccess>` is then merged with the previous in the `Vec<Vec<...>>`, or else re-appended to the empty `Vec<Vec<...>>`.

### For storage

Internally, `Forge` maintains a `Vec<Vec<RecordedStorageAccess>>`. The vector is 2-dimensional to allow for grouping storage accesses by relative call depth, so they can be updated with the eventual failure status of that callframe before being merged with the previous `Vec` to preserve absolute ordering.

On each `SLOAD` or `SSTORE`, a `RecordedStorageAccess` is appended to the last `Vec<RecordedStorageAccess>`, or else appended as a single-element `Vec`.

On `call_end` or `create_end`, the last `Vec<RecordedStorageAccess>` is `pop()`'d off and the accesses therein are updated with the failure status of the `CREATE` or `CALL` opcode. This `Vec<RecordedStorageAccess>` is then merged with the previous in the `Vec<Vec<...>>`, or else re-appended to the empty `Vec<Vec<...>>`

### Rationale

Ordering is important when manually calculating account or storage-slot "warm-ness," as is the eventual `reverted` status of a callframe in which an account or storage access happened. 

Opcodes have different costs depending on whether or not a storage slot is "warm," "cold," or going from zero-to-non-zero value, or when sending value to an empty/uninitialized account. [On an even lower-level, the EVM does not mark "cold" accounts or slots as "warm" if the context from which they were accessed reverts (with the exception of the counterfactual `create` address)](https://twitter.com/emo_eth/status/1696281639201259677).

Thus, context like `previousValue`, `newValue`, `isCreate,` `initialized`, and `reverted` are important for understanding and evaluating the (hypothetical) behavior of the EVM. 

## Open questions

- ~Should SELFDESTRUCT "accesses" be recorded? They have no effect on internal EVM state "warmness," but they may be difficult to track otherwise~
  - The opcode is deprecated/scheduled for removal for Ethereum networks, but no guarantees can be made for other chains or execution environments
  - Selfdestruct calls can vary in gas depending on whether or not an account is initialized, and there is no other way to access selfdestruct-targeted accounts, so I have added the ability to track them via the cheatcode.
- Should `AccountAccesses` track `CREATE2` salts? It may be difficult or burdensome to anticipate how a particular contract determines `CREATE2` salt

### CODE REVIEW

**Review by Evalir — CHANGES_REQUESTED** [review]
Still haven't properly tested functionality / architecture here, but looks good so far! a couple of comments to start jamming on this

File: testdata/foundry.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ invariant_shrink_sequence = true
 gas_limit = 9223372036854775807
 gas_price = 0
 gas_reports = ["*"]
-ignored_error_codes = [1878]
+ignored_error_codes = [1878, 5159]
```
[the remark this answers is not in the document]
Comment (jameswenzel) [reply testdata/foundry.toml:23]: This is the `SELFDESTRUCT` deprecation warning – one of the tests explicitly tests `SELFDESTRUCT` behavior so it is ignored. (I personally suspect alt-EVM chains will support SELFDESTRUCT for years to come)

File: testdata/foundry.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [profile.default]
-solc = "0.8.19"
+solc = "0.8.18"
```
[the remark this answers is not in the document]
Comment (jameswenzel) [reply testdata/foundry.toml:2]: Most of the tests specify a strict `0.8.18` version

**Comment by mds1** [comment]
Hey @jameswenzel @Evalir, aside from the conflicts are there any blockers or unanswered questions here?

**Comment by refcell** [comment]
A simpler way of doing this would be something like what's described in https://github.com/foundry-rs/foundry/issues/6125

This would make the api be a single cheatcode call and coalesce storage and account access recording to make state diff generation simpler and more extensible for tracing.

**Comment by jameswenzel** [comment]
Closing in favor of #6087. @refcell, love it!

**Pull request closed without merging**